src/tests/end2end/DepthStencilStateTests.cpp

```diff
@@ -0,0 +1,512 @@
+        .SetDepthWriteEnabled(false)
+        .GetResult();
+
+    nxt::DepthStencilState checkState = device.CreateDepthStencilStateBuilder()
+        .SetDepthCompareFunction(nxt::CompareFunction::Less)
+        .GetResult();
+
+    DoTest({
+        { baseState, RGBA8(255, 255, 255, 255), 1.f, 0u },
+        { noDepthWrite, RGBA8(0, 0, 0, 255), 0.f, 0u },
+        { checkState, RGBA8(0, 255, 0, 255), 0.5f, 0u },
+    }, RGBA8(0, 255, 0, 255));
+}
+
+TEST_P(DepthStencilStateTest, StencilAlways) {
+    CheckStencilCompareFunction(nxt::CompareFunction::Always, true, true, true);
+}
+
+TEST_P(DepthStencilStateTest, StencilEqual) {
+    CheckStencilCompareFunction(nxt::CompareFunction::Equal, false, true, false);
+}
+
+TEST_P(DepthStencilStateTest, StencilGreater) {
+    CheckStencilCompareFunction(nxt::CompareFunction::Greater, false, false, true);
+}
+
+TEST_P(DepthStencilStateTest, StencilGreaterEqual) {
+    CheckStencilCompareFunction(nxt::CompareFunction::GreaterEqual, false, true, true);
+}
+
+TEST_P(DepthStencilStateTest, StencilLess) {
+    CheckStencilCompareFunction(nxt::CompareFunction::Less, true, false, false);
+}
+
+TEST_P(DepthStencilStateTest, StencilLessEqual) {
+    CheckStencilCompareFunction(nxt::CompareFunction::LessEqual, true, true, false);
+}
+
+TEST_P(DepthStencilStateTest, StencilNever) {
+    CheckStencilCompareFunction(nxt::CompareFunction::Never, false, false, false);
+}
+
+TEST_P(DepthStencilStateTest, StencilNotEqual) {
+    CheckStencilCompareFunction(nxt::CompareFunction::NotEqual, true, false, true);
+}
+
+TEST_P(DepthStencilStateTest, StencilKeep) {
+    CheckStencilOperation(nxt::StencilOperation::Keep, 1, 3, 1);
+}
+
+TEST_P(DepthStencilStateTest, StencilZero) {
+    CheckStencilOperation(nxt::StencilOperation::Zero, 1, 3, 0);
+}
+
+TEST_P(DepthStencilStateTest, StencilReplace) {
+    CheckStencilOperation(nxt::StencilOperation::Replace, 1, 3, 3);
+}
+
+TEST_P(DepthStencilStateTest, StencilInvert) {
+    CheckStencilOperation(nxt::StencilOperation::Invert, 0xf0, 3, 0x0f);
+}
+
+TEST_P(DepthStencilStateTest, StencilIncrementClamp) {
+    CheckStencilOperation(nxt::StencilOperation::IncrementClamp, 1, 3, 2);
+    CheckStencilOperation(nxt::StencilOperation::IncrementClamp, 0xff, 3, 0xff);
+}
+
+TEST_P(DepthStencilStateTest, StencilIncrementWrap) {
+    CheckStencilOperation(nxt::StencilOperation::IncrementWrap, 1, 3, 2);
+    CheckStencilOperation(nxt::StencilOperation::IncrementWrap, 0xff, 3, 0);
+}
+
+TEST_P(DepthStencilStateTest, StencilDecrementClamp) {
+    CheckStencilOperation(nxt::StencilOperation::DecrementClamp, 1, 3, 0);
+    CheckStencilOperation(nxt::StencilOperation::DecrementClamp, 0, 3, 0);
+}
+
+TEST_P(DepthStencilStateTest, StencilDecrementWrap) {
+    CheckStencilOperation(nxt::StencilOperation::DecrementWrap, 1, 3, 0);
+    CheckStencilOperation(nxt::StencilOperation::DecrementWrap, 0, 3, 0xff);
+}
+
+TEST_P(DepthStencilStateTest, StencilReadMask) {
```

Comment:
nit: comment before test and a comment in the test explaining how it tests the read mask.